app/javascript/components/widget-controller.jsx

```diff
@@ -33,234 +29,58 @@ const getWidgets = gql`
   }
 `;
 
-const Wrapper = styled.div`
-  outline: none;
-`;
-
-const widgetElements = {
-  Weather,
-  Twitter,
-  Numbers,
-  Traffic,
-};
-
-export class WidgetController extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      currentWidgetId: 0,
-      transitionTime: null,
-    };
-  }
-
-  componentDidMount() {
-    const prefetchDelay = 1500;
+export const WidgetController = ({ client, cityName }) => {
+  const [newWidgetId, setNewWidgetId] = useState(0);
 
-    this.lastPrefetchTime = 0;
-    this.timerId = setInterval(() => {
-      const { transitionTime, pausedTime } = this.state;
-
-      if (!transitionTime || pausedTime) {
-        return;
-      }
-
-      const { client } = this.props;
-      const prefetchTime = transitionTime - prefetchDelay;
-      const now = Date.now();
-
-      if (transitionTime < now) {
-        this.nextWidget();
-      } else if (
-        prefetchTime < now &&
-        this.lastPrefetchTime < now - prefetchDelay
-      ) {
-        const { nextWidgetId } = this.state;
-        const { cityName } = this.props;
-        this.lastPrefetchTime = now;
-        client.query({
-          query: getWidgets,
-          variables: {
-            id: nextWidgetId,
-            cityName,
-          },
-        });
-      }
-    }, 500);
-  }
-
-  componentWillUnmount() {
-    if (this.timerId) {
-      clearInterval(this.timerId);
-    }
-  }
-
-  switchPages = ({ keyCode }) => {
-    const upKey = 38;
-    const downKey = 40;
-    const SPACEBAR = 32;
-    if (keyCode === SPACEBAR) {
-      this.setState(({ pausedTime, transitionTime }) =>
-        pausedTime
-          ? {
-              pausedTime: null,
-              transitionTime: transitionTime - pausedTime + Date.now(),
-            }
-          : { pausedTime: Date.now() },
-      );
-    }
-    if (keyCode === downKey) {
-      this.nextWidget();
-    }
-    if (keyCode === upKey) {
-      this.prevWidget();
-    }
-  };
+  const fetchPolicy = newWidgetId === 0 ? 'network-only' : 'cache-first';
```

Comment:
what does this line do ?

Reply:
I'm curious too - I believe this was a line added in tucker's branch? @tucker-weed

Reply:
The fetch policy is maintained from the previous version of WidgetController - my guess about what is going on here is that the first ever query run by the WidgetController (when `newWidgetId` is 0) should make a network request, whereas all subsequent queries should check if needed query information is in the cache before making a network request. Not 100% sure about this though so if anyone else knows what's really going on here, let us know.

Reply:
Filling in the missing piece. Its because we prefetch with a direct call to `client.query`. The flow:
* Page load / First request. `<Query`  uses network-only to get the latest data to make sure we're not stale. Updates cache for first widget.
* 1.5 seconds before transition, call `client.query`. This is always a network request. Updates cache for second widget.
* Transition time. Update state to next widget. `<Query` is not `cache-first`. `<Query` has a cache hit from the update from `client.query` 1.5 seconds ago.
* Repeat the previous two steps.

If we only use `network-only` without the prefetch, our transitions are delayed by the server response time. This is compounded by also loading assets after so it was visibly slow.
If we only use `cache-first` our very first request will be our last widget from page close with old time/weather data.